llpc/context/llpcCompiler.cpp

```diff
@@ -728,6 +728,246 @@ template <typename T> static FormattedBytes formatBytesLittleEndian(ArrayRef<T>
+    context->attachPipelineContext(&graphicsContext);
+
+    result = buildUnlinkedShaderInternal(context, shaderInfo, stage, candidateElf, pipelineOut->stageCacheAccesses);
+    releaseContext(context);
+
+    if (result == Result::Success) {
+      elfBin.codeSize = candidateElf.size();
+      elfBin.pCode = candidateElf.data();
+    }
+
+    if (cacheAccessor && pipelineOut->pipelineCacheAccess == CacheAccessInfo::CacheNotChecked)
+      pipelineOut->pipelineCacheAccess = CacheAccessInfo::CacheMiss;
+  } else {
+    LLPC_OUTS("Cache hit for graphics pipeline.\n");
+    elfBin = cacheAccessor->getElfFromCache();
+    if (cacheAccessor->isInCache()) {
+      pipelineOut->pipelineCacheAccess =
+          cacheAccessor->hitInternalCache() ? CacheAccessInfo::InternalCacheHit : CacheAccessInfo::CacheHit;
+    }
+  }
+
+  if (result == Result::Success) {
+    void *allocBuf = nullptr;
+    if (pipelineInfo->pfnOutputAlloc) {
+      allocBuf = pipelineInfo->pfnOutputAlloc(pipelineInfo->pInstance, pipelineInfo->pUserData, elfBin.codeSize);
+
+      uint8_t *code = static_cast<uint8_t *>(allocBuf);
+      memcpy(code, elfBin.pCode, elfBin.codeSize);
+
+      pipelineOut->pipelineBin.codeSize = elfBin.codeSize;
+      pipelineOut->pipelineBin.pCode = code;
+    } else {
+      // Allocator is not specified
+      result = Result::ErrorInvalidPointer;
+    }
+  }
+
+  if (cacheAccessor && !cacheAccessor->isInCache() && result == Result::Success) {
+    LLPC_OUTS("Adding graphics pipeline to the cache.\n");
+    cacheAccessor->setElfInCache(elfBin);
+  }
+  return result;
+}
+
+// Build shader from the specified info.
+//
+// @param [in]  pipelineInfo    Info to build this shader module
+// @param [out] pipelineOut  : Output of building this shader module
+// @param [in]  pElfpackage :  Early compiled elfPackage
+//
+// @returns : Result::Success if successful. Other return codes indicate failure.
+Result Compiler::buildGraphicsPipelineWithElf(const GraphicsPipelineBuildInfo *pipelineInfo,
+                                              GraphicsPipelineBuildOut *pipelineOut, const BinaryData *pElfpackage) {
+
+  Result result = Result::Success;
+  // clang-format off
+  SmallVector<const PipelineShaderInfo *, ShaderStageGfxCount> shaderInfo = {
+    &pipelineInfo->vs,
+    &pipelineInfo->tcs,
+    &pipelineInfo->tes,
+    &pipelineInfo->gs,
+    &pipelineInfo->fs,
+  };
+  // clang-format on
+
+  ElfPackage elf[UnlinkedStageCount] = {};
+
+  MetroHash::Hash cacheHash = {};
+  MetroHash::Hash pipelineHash = {};
+  cacheHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+  pipelineHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+
+  GraphicsContext graphicsContext(m_gfxIp, pipelineInfo, &pipelineHash, &cacheHash);
+  Context *context = acquireContext();
+  context->attachPipelineContext(&graphicsContext);
+
+  for (unsigned idx = 0; idx < UnlinkedStageCount; idx++) {
```

Comment:
Can you iterate over all unlinked stages instead (with `enumRange`)?

Reply:
Yes, but index is also needed.